### PR TITLE
Enable frozen string literal

### PR DESCRIPTION
**Pull request by aliismayilov**
https://github.com/haml/haml/pull/967 added the frozen string literal magic comments on all ruby files.

To ensure that the strings are not modified in the future, I think the tests also should cover it. This PR fixes a string mutation case and runs tests with `RUBYOPT=--enable-frozen-string-literal`, where the ruby interpreter will ensure the frozen strings.

### CODE REVIEW

**Comment by aliismayilov** [comment]
I'm not sure how to deal with jruby warning causing the tests to fail. Any input would be welcome

**Comment by hcatlin** [comment]
Ooff, that's annoying.

File: lib/haml/util.rb [hunk 1]
```diff
@@ -213,7 +213,7 @@ def unescape_interpolation(str, escape_html = nil)
             scan.scan(/\w+/)
           end
           content = eval("\"#{interpolated}\"")
-          content.prepend(char) if char == '@' || char == '$'
+          content = "#{char}#{content}" if char == '@' || char == '$'
```
Comment (hcatlin) [inline lib/haml/util.rb:216]: I'd be curious if there are any performance issues here...?

File: test/cases/engine_internals_test.rb [hunk 2]
```diff
@@ -37,7 +37,7 @@ def inspect
         print Haml::Engine.new('%div{ foo: true, bar: false }').render
       HAML
       f.close
-      out = IO.popen([RbConfig.ruby, f.path], &:read)
+      out = IO.popen([RbConfig.ruby, '-W0', f.path], &:read)
```
Comment (hcatlin) [inline test/cases/engine_internals_test.rb:40]: Can you add a documentation note here on why the `-W0` is there and what it's used for?

File: test/cases/nesting_test.rb [hunk 3]
```diff
@@ -160,7 +160,7 @@ def test_silent_end_with_stuff
 b
 a
 HTML
-- str = "abcde"
+- str = +"abcde"
```
Comment (hcatlin) [inline test/cases/nesting_test.rb:163]: Wow, I had no idea about this feature... it took me some serious digging to learn what it even was. Great work!

**Comment by hcatlin** [comment]
Ran the tests locally with the same version of JRuby and it passed just fine... still investiating.

**Comment by hcatlin** [comment]
Since the intent of the breaking test is to capture an interesting edge case, I have slightly widened it's expectations, so that we don't have to be 100% sure that ALL interpreters NEVER print a warning when executed as a child or in various different configurations.

**Comment by hcatlin** [comment]
Further, this fails on the `main` branch: https://travis-ci.org/github/haml/haml/jobs/731023617

File: test/cases/engine_internals_test.rb (hunk 2 above)
Comment (aliismayilov) [reply test/cases/engine_internals_test.rb:44]: here the I attempted to set ruby interpreter warning level to 0 (=silence), with hopes that jruby warning wouldn't be printed out. I see that you've already merged the PR, I guess, I can leave it as is.
